BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m

```diff
@@ -249,19 +249,28 @@ - (void)failAuthenticationWithURLSessionTask:(NSURLSessionTask *)task
     [self setWebViewCanBeUsedDirectly:NO forHost:task.currentRequest.URL.host];
 
     // We can only handle the connection error if we're in the view hierarchy. It would be difficult to display UIAlertControllers when not sure we're in the view hierarchy. We will try again after viewDidAppear
-    if (self.view.window) {
-        [self handleConnectionErrorWithError:error message:message];
-    } else {
-        self.connectionError = error;
-        self.connectionErrorMessage = message;
-        // We still want to report the completion block so that the necessary cleanup is done up in the call tree.
-        [self prepareForDismissal];
-        if (self.completionBlock) {
-            self.completionBlock(self, nil, self.connectionError);
-        }
-    }
+    BOOL isMainThread = [NSThread isMainThread];
+    dispatch_async(dispatch_get_main_queue(), ^{
+        //dispatch on main thread UI-related activities
+        if (self.view.window) {
+            [self handleConnectionErrorWithError:error message:message];
+            [self.activityIndicator stopAnimating];
+        } else {
+            [BOXDispatchHelper callCompletionBlock:^{
```

Comment:
I don't think there's a need to pick up the dispatch helper here and create a new dependency. There should be no expectation on what thread 'prepareForDismissal' or the completion block get called on. They can be on the main thread and it would be the responsibility of the implementations of those to dispatch high-intensity work off to some worker queue.